File: lce_p0242_valid_anagram.java
/*
  LCE 242. Valid Anagram

  Given two strings s and t, return true if t is an anagram of s, and false otherwise.

  Constraints:
  - 1 <= s.length, t.length <= 5 * 10^4
  - s and t consist of lowercase English letters.

  Topics: 
  - Hash Table
  - String
  - Sorting
*/

import java.util.Arrays;

class Solution {

  // Time Complexity: O(n) - 2 ms -> 97.88%
  // Space Complexity: O(1) - 43.0 MB -> 76.06%
  public boolean isAnagram(String s, String t) {

    if (s.length() != t.length()) {
      return false;
    }

    int[] freqS = new int[26];
    int[] freqT = new int[26];

    for (char ch : s.toCharArray()) {
      freqS[ch - 'a']++;
    }
    for (char ch : t.toCharArray()) {
      freqT[ch - 'a']++;
    }

    return Arrays.equals(freqS, freqT);
  }

  // Time Complexity: O(n) - 6 ms -> 44.40%
  // Space Complexity: O(1) - 42.9 MB -> 85.96%
  public boolean isAnagramAlt1(String s, String t) {

    if (s.length() != t.length()) {
      return false;
    }

    int[] freqArr = new int[26];

    for (int i = 0; i < s.length(); i++) {
      freqArr[s.charAt(i) - 'a']++;
      freqArr[t.charAt(i) - 'a']--;
    }

    for (int freq : freqArr) {
      if (freq != 0) {
        return false;
      }
    }

    return true;
  }

  // Time Complexity: O(n log n + m log m) - 4 ms -> 76.24%
  // Space Complexity: O(1) - 44.9 MB -> 13.69%
  public boolean isAnagramAlt2(String s, String t) {

    if (s.length() != t.length()) {
      return false;
    }

    char[] charS = s.toCharArray();
    char[] charT = t.toCharArray();
    Arrays.sort(charS);
    Arrays.sort(charT);
    
    return Arrays.equals(charS, charT);
  }
}

/*
  methods:
  1. frequency array (compares array contents)
  2. frequency array (checks for equilibrium)
  3. sorting (sort then compare array contents)
*/

File: python/lce_p0242_valid_anagram.py
"""
LCE 242. Valid Anagram

Given two strings s and t, return true if t is an anagram of s, and false otherwise.

Constraints:
- 1 <= s.length, t.length <= 5 * 104
- s and t consist of lowercase English letters.

Topics:
- Hash Table
- String
- Sorting
"""


class Solution:

    # Time Complexity: O(n) - 7 ms -> 89.75%
    # Space Complexity: O(n) - 18.12 MB -> 25.21%
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False

        return Counter(s) == Counter(t)

    # Time Complexity: O(n) - 11 ms -> 72.85%
    # Space Complexity: O(n) - 17.65 MB -> 97.43%
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False

        freq_s = defaultdict(int)
        freq_t = defaultdict(int)

        for i in range(len(s)):
            freq_s[s[i]] += 1
            freq_t[t[i]] += 1

        return freq_s == freq_t

    # Time Complexity: O(n) - 11 ms -> 72.85%
    # Space Complexity: O(n) - 17.86 MB -> 64.17%
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False

        n = len(s)
        freq_arr = [0] * 26

        for i in range(n):
            freq_arr[ord(s[i]) - ord("a")] += 1
            freq_arr[ord(t[i]) - ord("a")] -= 1

        for freq in freq_arr:
            if freq != 0:
                return False

        return True

    # Time Complexity: O(n log n) - 16 ms -> 26.19%
    # Space Complexity: O(n) - 18.46 MB -> 19.55%
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False

        return sorted(s) == sorted(t)


"""
  methods:
  1. Counter
  2. freq defaultdict
  3. freq array
  4. sorting
"""
